---
title: "Ros_paper"
output: html_document
date: '2023-02-16'
---
# Copyright (C) 2023 University of Cambridge
This is a script by Jens Bager Christensen (jbc53@cam.ac.uk) to perform pseudobulk analysis on HOPX-NEPs to identify their transcriptional response to neonatal injury.

```{r}
#Load packages
library(Seurat)
library(tidyverse)
library(DESeq2)
library(EnhancedVolcano)
library(goseq)

#Defining a function to test of different resolutions for clustering.
res.range <- c(seq(0.1,2,0.1))
assay <- "SCT"
dims <- 20
reduction <- "pca"

#Defining a function, which clusters a RDS in a range of resolutions, and calculates both the silhouette score and the negative silhouette proportion to determine which resolution is optimal.
clusteringKit <- function(name, assay, dim, res.range, reduction){
  DefaultAssay(name) <- assay
  #using FindNeighbors (Seurat), a KNN based clustering algorithm, to compute a SNN graph, which then can be used to cluster the cells w. their neighbors.
  name <- FindNeighbors(name, dims=1:dim, reduction=reduction)
  for (i in res.range){
    #FindClusters is used to cluster the aforementioned SNN graph by optimising the modularity function. This is carried out for all resolutions in the range of interest.
    name <- FindClusters(name, resolution=i)
  }
DefaultAssay(name) <- assay
  dist.matrix <- dist(x = Embeddings(object = name[[reduction]])[, 1:dim])
 clusters <- paste0(assay,"_snn_res.", res.range)
  getSil <- function(clr) {
    clr <- name@meta.data[[clr]]
    sil <- cluster::silhouette(x = as.numeric(x = as.factor(x = clr)), dist = dist.matrix)
    sil_value <- sil
    return(sil_value)
  }
  
  sls <- lapply(clusters, getSil) #mc.cores = 1)
  sls_median <- sapply(sls, function(x) median(x[,3])) %>% setNames(., res.range)
  sls_neg_prop <- sapply(sls, function(x) sum(x[,3]<0)/length(x[,3])) %>% setNames(., res.range)
  p_list <- lapply(res.range, function(res){
    Idents(name) = name@meta.data[paste0(assay, "_snn_res.",res)]
    DimPlot(name, reduction = "umap", label = TRUE, group.by = paste0(assay, "_snn_res.",res), pt.size =0.3, raster=F)
  })
  return(list(name=name, sls_median=sls_median, sls_neg_prop=sls_neg_prop, dimplots=p_list))
}
```


#This is the script for performing DESeq2 analysis on Hopx positive NEPs to illustrate their response to injury.

The RDS object is read in, which previously has been preprocessed and mapped using CellRanger.
```{r}
#read in the RDS object obtained was described in Pakula et al., 2023.
data <- readRDS(file = "~/../jbc53/data/nep.rna.combined.solo.RDS")

#add cluster annotations to the metadata.
data@meta.data <- data@meta.data %>%
  mutate(cluster = Idents(data))

#alter misspelling of a cluster name.
data@meta.data$cluster <- recode_factor(data@meta.data$cluster, "Ependymal_!" = "Ependymal_1")

#inspect the dimplot.
DimPlot(data, label = T, label.box = T, repel = T, group.by = "cluster") & NoLegend() & NoAxes()

#Create a DotPlot of key markers to explain the annotations.
DotPlot(data, group.by = "seurat_clusters", features = c("Hopx", "Gdf10", "Slc1a3", "Ascl1", "Ptf1a", "Tubb3", "Atoh1", "Pax2", "Mki67", "Top2a", "Cx3cr1", "Olig2", "Gpr17", "Foxj1", "Vtn"))
```

Visualising expression changing in time in Hopx positive clusters
```{r}
#setting a seed, ensuring consistent clustering.
set.seed(123)

#Inspect Hopx expression in a featureplot and a violin plot to determine which clusters to include.
FeaturePlot(data, features = "Hopx") & NoAxes()
VlnPlot(data, features = "Hopx", group.by = "cluster", sort = "increasing")
#clusters BgL-NEP1, -2, -3, and -4 are deemed Hopx positive and thus used for further analysis.

#create the subset containing all Hopx positive NEP clusters, excluding the ones from P1.
Idents(data) <- data@meta.data$cluster
hopx_all <- subset(data, idents = c("BgL_NEP_1","BgL_NEP_2","BgL_NEP_3","BgL_NEP_4")) %>% subset(timepoint != "P1")

#plot expression of markers of interesting changing with respect to time Reordering the levels to make nonIR first.
hopx_all@meta.data$treatment <- factor(hopx_all@meta.data$treatment, levels = c("nonIR", "IR"))
VlnPlot(hopx_all, features = "Cdkn1a", group.by = "timepoint", split.by = "treatment")
VlnPlot(hopx_all, features = "Phlda3", group.by = "timepoint", split.by = "treatment")
VlnPlot(hopx_all, features = "Ass1", group.by = "timepoint", split.by = "treatment")
VlnPlot(hopx_all, features = "Bax", group.by = "timepoint", split.by = "treatment")
VlnPlot(hopx_all, features = "Ano3", group.by = "timepoint", split.by = "treatment")
VlnPlot(hopx_all, features = "Fas", group.by = "timepoint", split.by = "treatment")
```

We're interested in the Hopx expressing NEPs. Clusters identified as HOPX-NEPs are subsetted. To divide the initial, acute injury response with later biological reactions, the data is split into P2 cells and P3+P5 cells. Pseudobulk analysis is performed on these two data set separately.   
```{r}
#setting a seed, ensuring consistent clustering.
set.seed(123)

#Inspect Hopx expression in a featureplot and a violin plot to determine which clusters to include.
FeaturePlot(data, features = "Hopx") & NoAxes()
VlnPlot(data, features = "Hopx", group.by = "cluster", sort = "increasing")
#clusters BgL-NEP1, -2, -3, and -4 are deemed Hopx positive and thus used for further analysis

#subset the HOPX-NEPs from P2 only.
hopx_p2 <- subset(data, idents = c("BgL_NEP_1","BgL_NEP_2","BgL_NEP_3","BgL_NEP_4")) %>% subset(timepoint == "P2")

#split the object allowing for reintegration.
hopx_p2_list <- SplitObject(hopx_p2, split.by = "orig.ident")


#Re-integrate datasets.
hopx_p2_list <- lapply(X = hopx_p2_list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 3000)
})


#select features that are repeatedly variable across datasets for integration.
features.hopx_p2 <- SelectIntegrationFeatures(object.list = hopx_p2_list)
rna.anchors.hopx_p2 <- FindIntegrationAnchors(object.list = hopx_p2_list, anchor.features = features.hopx_p2)

#this command creates an 'integrated' data assay.
hopx_p2 <- IntegrateData(anchorset = rna.anchors.hopx_p2)

#regress out CC, mitrochondrial -, gene reads, and sequencing depth.
hopx_p2 <- SCTransform(hopx_p2, vars.to.regress = c("percent.mt", "CC.Difference", "nFeature_RNA", "nCount_RNA"))

#run PCA and UMAP w. parameters previously used.
hopx_p2 <- RunPCA(hopx_p2, features = hopx_p2@assays$integrated@var.features, npcs = 100)
hopx_p2 <- RunUMAP(hopx_p2, dims = 1:40, n.neighbours = 30L, seed.use = 888, repulsion.strength = 0.1, min.dist = 0.5, n.epochs = NULL)

#Visualise the data.
DimPlot(hopx_p2, label = T, label.box = T, repel = T) + ggtitle("UMAP of subsetted NEP types") & NoLegend() & NoAxes()

#Recluster the data trying out different clustering resolutions to figure out an optimal resolution.
hopx_p2_combined <- clusteringKit(hopx_p2, assay="SCT", dim=30, res.range=seq(0.1, 3, 0.1), reduction="pca")

#plot to find the highest silhouette score w. low negative silhouette proportion.
plot(hopx_p2_combined$sls_median~seq(0.1,3,0.1), type = "l", ylab = "Silhouette score") & plot(hopx_p2_combined$sls_neg_prop~seq(0.1,3,0.1), type = "l", ylab = "Negative silhouette proportion score")
#optimal resolution is seemingly around 0.7 - 0.8. The resolutions are evaluated for biological relevance to determine the actual optimal resolution.

#rename the Seurat objects.
hopx_p2 <- hopx_p2_combined$name

#Plot with the new resolutions.
DimPlot(hopx_p2, label = T, label.box = T, repel = T, group.by = "SCT_snn_res.0.8") + ggtitle("Cells from Hopx positive clusters from P2") & NoLegend() & NoAxes()
DimPlot(hopx_p2, label = F, label.box = T, repel = T, group.by = "timepoint", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P2") & NoAxes()
DimPlot(hopx_p2, label = F, label.box = T, repel = T, group.by = "treatment", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P2") & NoAxes()
DimPlot(hopx_p2, label = F, label.box = T, repel = T, group.by = "cluster", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P2") & NoAxes()

#See the Hopx expression in the subset.
FeaturePlot(hopx_p2, features = "Hopx") & NoAxes()
VlnPlot(hopx_p2, features = "Hopx", group.by = "cluster", sort = "increasing")
```


Performing the pseudobulk DEG analysis on only cells from P2, trying to capture the injury signal. Additionally, this matches the timepoint of the other experiments.
```{r}
#Choose the RDS object containing the cell of interest.
Hopx_p2_deseq2 <- hopx_p2 

#make a sample column in the meta data.
Hopx_p2_deseq2@meta.data$samples <- paste0(Hopx_p2_deseq2$orig.ident)

#Aggregate the count matrixes from the RDS object.
Hopx_p2_cts <- AggregateExpression(Hopx_p2_deseq2, group.by = c("samples", "treatment"),
                           assays = "RNA",
                           slot = "counts",
                           return.seurat = F)

#ensure the active assay is the RNA assay.
Hopx_p2_cts <- Hopx_p2_cts$RNA

#create a tibble containing the meta data necessary for performing the correct comparisons.
Hopx_p2_colData <- tibble(samples = colnames(Hopx_p2_cts))
Hopx_p2_colData <- Hopx_p2_colData %>% mutate(condition = ifelse(grepl('non', samples), 'nonIR', 'IR')) %>% column_to_rownames(var = 'samples')

#create an object suitable for a DESeq2 analysis.
hopx_p2_dds <- DESeqDataSetFromMatrix(countData = Hopx_p2_cts,
                              colData = Hopx_p2_colData,
                              design = ~ condition)

#remove too lowly expressed genes.
hopx_p2_keep <- rowSums(counts(hopx_p2_dds)) >= 10
hopx_p2_dds <- hopx_p2_dds[hopx_p2_keep,]

#perform the actual DESeq2 analysis.
hopx_p2_dds <- DESeq(hopx_p2_dds)

#identify the results of interest and save it in a variable.
resultsNames(hopx_p2_dds)
hopx_p2_res <- results(hopx_p2_dds, name = "condition_nonIR_vs_IR")

#visualise the results of the DESeq2 analysis in a volano plot.
EnhancedVolcano(hopx_p2_res, lab = row.names(hopx_p2_res), x = 'log2FoldChange', y = 'padj', pCutoff = 0.05, FCcutoff = 0.5, labSize = 4, cutoffLineType = "blank", xlab = "Log fold-change", ylab = "-Log adjusted p-val" ,   colAlpha = 1, legendPosition = 'right', legendLabSize = 12, legendIconSize = 4.0, drawConnectors = T) & NoLegend()

#save csv of the DESeq2 analysis
#write.csv(hopx_p2_res, file = "Hopx/Hopx_p2_DESeq2.csv")
```


Performing GO term analysis to see, if any ROS related pathways are upregulated in the IR cells.
```{r}
#restructuring the DESeq2 results making it fit for GO term analysis.
hopx_p2_res_go <- hopx_p2_res %>% as.tibble() %>% transform(gene = row.names(hopx_p2_res))

#identify the genes which are significantly differentially expressed in the IR cells.
hopx_p2_isSigGene <- hopx_p2_res_go$padj < 0.05 & !is.na(hopx_p2_res_go$padj) & hopx_p2_res_go$log2FoldChange < -0.5
hopx_p2_genes <- as.integer(hopx_p2_isSigGene)
names(hopx_p2_genes) <- hopx_p2_res_go$gene

#calculate a probability weight function for the significantly DE genes in IR cells.
hopx_p2_pwf <- nullp(hopx_p2_genes, "mm8", "geneSymbol")

#perform the goseq analysis.
hopx_p2_goResults <- goseq(hopx_p2_pwf, "mm8","geneSymbol", test.cats=c("GO:BP"))

#visualise the top 10 overrepresented GO terms.
hopx_p2_goResults %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for IR cells from P2 of Hopx positive clusters") + theme_bw()


#visualisation of selected GO terms.
hopx_p2_goResults %>% 
    subset(category %in% c("GO:0072331", "GO:0006915", "GO:0006950", "GO:0008630", "GO:0042771", "GO:0071478", "GO:0006974", "GO:0045926", "GO:0071479", "GO:0031571", "GO:0072593")) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for IR cells from P2 of Hopx positive clusters") + theme_bw()

#look a stress-related GO terms to see if any of these are significantly overrepresented.
hopx_p2_goResults[grep("stress", hopx_p2_goResults$term), ]

#Write a csv of the GO term results.
#write.csv(hopx_p2_goResults, file = "Hopx/Hopx_P2_GO_term.csv")
```

Performing the GO term analysis on the DEGs upregulated in nonIR cells.
```{r}
#restructuring the DESeq2 results making it fit for GO term analysis.
hopx_p2_res_go <- hopx_p2_res %>% as.tibble() %>% transform(gene = row.names(hopx_p2_res))

#identify the genes which are significantly differentially expressed in the nonIR cells.
hopx_p2_isSigGene_nonIR <- hopx_p2_res_go$padj < 0.05 & !is.na(hopx_p2_res_go$padj) & hopx_p2_res_go$log2FoldChange > 0.5
hopx_p2_genes_nonIR <- as.integer(hopx_p2_isSigGene_nonIR)
names(hopx_p2_genes_nonIR) <- hopx_p2_res_go$gene

#calculate a probability weight function for the significantly DE genes in nonIR cells.
hopx_p2_pwf_nonIR <- nullp(hopx_p2_genes_nonIR, "mm8", "geneSymbol")

#perform the goseq analysis.
hopx_p2_goResults_nonIR <- goseq(hopx_p2_pwf_nonIR, "mm8","geneSymbol", test.cats=c("GO:BP"))

#visualise the top 10 overrepresented GO terms.
hopx_p2_goResults_nonIR %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for nonIR cells from P2 of Hopx positive clusters") + theme_bw()

#visualisation of selected GO terms.
hopx_p2_goResults_nonIR %>% 
    subset(category %in% c("GO:0006066", "GO:0006694", "GO:0006695", "GO:0006629", "GO:0044255", "GO:0006644")) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for nonIR cells from P2 of Hopx positive clusters") + theme_bw()

#look a stress-related GO terms to see if any of these are significantly overrepresented.
hopx_p2_goResults_nonIR[grep("stress", hopx_p2_goResults_nonIR$term), ]

#Write a csv of the GO term results.
#write.csv(hopx_p2_goResults_nonIR, file = "Hopx/Hopx_P2_GO_term_nonIR.csv")
```

Doing the analysis using P3 + P5 cells to compare early and later injury response.
```{r}
#setting a seed, ensuring consistent clustering.
set.seed(123)

#subset the HOPX-NEPs which are not from P1 or P2.
hopx_late <- subset(data, idents = c("BgL_NEP_1","BgL_NEP_2","BgL_NEP_3","BgL_NEP_4"))
hopx_late <- subset(hopx_late, timepoint != c("P1"))
hopx_late <- subset(hopx_late, timepoint != c("P2"))



#split the object allowing for reintegration.
hopx_late_list <- SplitObject(hopx_late, split.by = "orig.ident")


#Re-integrate datasets.
hopx_late_list <- lapply(X = hopx_late_list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 3000)
})


#select features that are repeatedly variable across datasets for integration.
features.hopx_late <- SelectIntegrationFeatures(object.list = hopx_late_list)
rna.anchors.hopx_late <- FindIntegrationAnchors(object.list = hopx_late_list, anchor.features = features.hopx_late)

#this command creates an 'integrated' data assay.
hopx_late <- IntegrateData(anchorset = rna.anchors.hopx_late)

#regress out CC, mitrochondrial -, gene reads, and sequencing depth.
hopx_late <- SCTransform(hopx_late, vars.to.regress = c("percent.mt", "CC.Difference", "nFeature_RNA", "nCount_RNA"))

#run PCA and UMAP w. parameters previously used.
hopx_late <- RunPCA(hopx_late, features = hopx_late@assays$integrated@var.features, npcs = 100)
hopx_late <- RunUMAP(hopx_late, dims = 1:40, n.neighbours = 30L, seed.use = 888, repulsion.strength = 0.1, min.dist = 0.5, n.epochs = NULL)

#Visualise the data.
DimPlot(hopx_late, label = T, label.box = T, repel = T) + ggtitle("UMAP of subsetted NEP types") & NoLegend() & NoAxes()

#Recluster the data trying out different clustering resolutions to figure out an optimal resolution.
hopx_late_combined <- clusteringKit(hopx_late, assay="SCT", dim=30, res.range=seq(0.1, 3, 0.1), reduction="pca")

#plot to find the higest silhouette score w. low negative silhouette proportion.
plot(hopx_late_combined$sls_median~seq(0.1,3,0.1), type = "l", ylab = "Silhouette score") & plot(hopx_late_combined$sls_neg_prop~seq(0.1,3,0.1), type = "l", ylab = "Negative silhouette proportion score")
#optimal resolution is seemingly around 0.7 - 0.8. The resolutions are evaluated for biological relevance to determine the actual optimal resolution.

#rename the Seurat object.
hopx_late <- hopx_late_combined$name

#Plot with the new resolutions.
DimPlot(hopx_late, label = T, label.box = T, repel = T, group.by = "SCT_snn_res.0.8") + ggtitle("Cells from Hopx positive clusters from P3 and P5") & NoLegend() & NoAxes()
DimPlot(hopx_late, label = F, label.box = T, repel = T, group.by = "timepoint", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P3 and P5") & NoAxes()
DimPlot(hopx_late, label = F, label.box = T, repel = T, group.by = "treatment", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P3 and P5") & NoAxes()
DimPlot(hopx_late, label = F, label.box = T, repel = T, group.by = "cluster", pt.size = 2) + ggtitle("Cells from Hopx positive clusters from P3 and P5") & NoAxes()

#Inspect Hopx expression in a featureplot and a violin plot to determine which clusters to include.
FeaturePlot(data, features = "Hopx") & NoAxes()
VlnPlot(data, features = "Hopx", group.by = "cluster", sort = "increasing") & NoLegend()
FeaturePlot(hopx_late, features = "Hopx") & NoAxes()
VlnPlot(hopx_late, features = "Hopx", group.by = "cluster", sort = "increasing")
#clusters BgL-NEP1, -2, -3, and -4 are deemed Hopx positive and thus used for further analysis.
```


Performing the pseudobulk DEG analysis on cells not from P2, trying to capture the injury signal. Additionally, this matches the timepoint of the other experiments.
```{r}
#Choose the RDS object containing the cell of interest.
hopx_late_deseq2 <- hopx_late 

#make a sample column in the meta data.
hopx_late_deseq2@meta.data$samples <- paste0(hopx_late_deseq2$orig.ident)

#Aggregate the count matrices from the RDS object.
hopx_late_cts <- AggregateExpression(hopx_late_deseq2, group.by = c("samples", "treatment"),
                           assays = "RNA",
                           slot = "counts",
                           return.seurat = F)

#ensure the active assay is the RNA assay.
hopx_late_cts <- hopx_late_cts$RNA

#create a tibble containing the meta data necessary for performing the correct comparisons.
hopx_late_colData <- tibble(samples = colnames(hopx_late_cts))
hopx_late_colData <- hopx_late_colData %>% mutate(condition = ifelse(grepl('non', samples), 'nonIR', 'IR')) %>% column_to_rownames(var = 'samples')

#create an object suitable for a DESeq2 analysis.
hopx_late_dds <- DESeqDataSetFromMatrix(countData = hopx_late_cts,
                              colData = hopx_late_colData,
                              design = ~ condition)

#remove too lowly expressed genes.
hopx_late_keep <- rowSums(counts(hopx_late_dds)) >= 10
hopx_late_dds <- hopx_late_dds[hopx_late_keep,]

#perform the actual DESeq2 analysis.
hopx_late_dds <- DESeq(hopx_late_dds)

#identify the results of interest and save it in a variable.
resultsNames(hopx_late_dds)
hopx_late_res <- results(hopx_late_dds, name = "condition_nonIR_vs_IR")

#visualise the results of the DESeq2 analysis in a volcano plot.
EnhancedVolcano(hopx_late_res, lab = row.names(hopx_late_res), x = 'log2FoldChange', y = 'padj', pCutoff = 0.05, FCcutoff = 0.5, labSize = 4, cutoffLineType = "blank", xlab = "Log fold-change", ylab = "-Log adjusted p-val" ,   colAlpha = 1, legendPosition = 'right', legendLabSize = 12, legendIconSize = 4.0, drawConnectors = T) & NoLegend()

#save csv of the DESeq2 analysis.
#write.csv(hopx_late_res, file = "Hopx/Hopx_P3_P5_DESeq2.csv")
```


Performing GO term analysis to see, if any ROS related pathways are upregulated in the IR cells
```{r}
#restructuring the DESeq2 results making it fit for GO term analysis.
hopx_late_res_go <- hopx_late_res %>% as.tibble() %>% transform(gene = row.names(hopx_late_res))

#identify the genes which are significantly differentially expressed in the IR cells.
hopx_late_isSigGene <- hopx_late_res_go$padj < 0.05 & !is.na(hopx_late_res_go$padj) & hopx_late_res_go$log2FoldChange < -0.5
hopx_late_genes <- as.integer(hopx_late_isSigGene)
names(hopx_late_genes) <- hopx_late_res_go$gene

#calculate a probability weight function for the significantly DE genes in IR cells.
hopx_late_pwf <- nullp(hopx_late_genes, "mm8", "geneSymbol")

#perform the goseq analysis.
hopx_late_goResults <- goseq(hopx_late_pwf, "mm8","geneSymbol", test.cats=c("GO:BP"))

#visualise the top 10 overrepresented GO terms.
hopx_late_goResults %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for IR cells from P3 and P5 of Hopx positive clusters") + theme_bw()

#visualisation of selected GO terms.
hopx_late_goResults %>% 
    subset(category %in% c("GO:0072332", "GO:0008630", "GO:0045429", "GO:0010212", "GO:0007095", "GO:0001558")) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for IR cells from P3 and P5 of Hopx positive clusters") + theme_bw()

#look a stress-related GO terms to see if any of these are significantly overrepresented.
hopx_late_goResults[grep("stress", hopx_late_goResults$term), ]

#Write a csv of the GO term results.
write.csv(hopx_late_goResults, file = "Hopx/Hopx_P3_P5_GO_term.csv")
```


Performing the same GO term analysis for DEGs upregulated in the nonIR cells.
```{r}
#restructuring the DESeq2 results making it fit for GO term analysis.
hopx_late_res_go <- hopx_late_res %>% as.tibble() %>% transform(gene = row.names(hopx_late_res))

#identify the genes which are significantly differentially expressed in the nonIR cells.
hopx_late_isSigGene_nonIR <- hopx_late_res_go$padj < 0.05 & !is.na(hopx_late_res_go$padj) & hopx_late_res_go$log2FoldChange > 0.5
hopx_late_genes_nonIR <- as.integer(hopx_late_isSigGene_nonIR)
names(hopx_late_genes_nonIR) <- hopx_late_res_go$gene

#calculate a probability weight function for the significantly DE genes in nonIR cells.
hopx_late_pwf_nonIR <- nullp(hopx_late_genes_nonIR, "mm8", "geneSymbol")

#perform the goseq analysis.
hopx_late_goResults_nonIR <- goseq(hopx_late_pwf_nonIR, "mm8","geneSymbol", test.cats=c("GO:BP"))

#visualise the top 10 overrepresented GO terms.
hopx_late_goResults_nonIR %>% 
    top_n(10, wt=-over_represented_pvalue) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for nonIR cells from P3 and P5 of Hopx positive clusters") + theme_bw()

#visualisation of selected GO terms.
hopx_late_goResults_nonIR %>% 
    subset(category %in% c("GO:0006650", "GO:0044255", "GO:0006629", "GO:0035788")) %>% 
    mutate(hitsPerc=numDEInCat*100/numInCat) %>% 
    ggplot(aes(x=hitsPerc, 
               y=term, 
               colour=over_represented_pvalue, 
               size=numDEInCat)) +
        geom_point() +
        expand_limits(x=0) +
        labs(x="Hits (%)", y="GO term", colour="p value", size="Count", title = "GO term analysis for nonIR cells from P3 and P5 of Hopx positive clusters") + theme_bw()

#look a stress-related GO terms to see if any of these are significantly overrepresented.
hopx_late_goResults_nonIR[grep("stress", hopx_late_goResults_nonIR$term), ]

#Write a csv of the GO term results.
#write.csv(hopx_late_goResults_nonIR, file = "Hopx/Hopx_P3_P5_GO_term_nonIR.csv")
```

